pkg/protocol/tipmanager/v1/tipmanager.go

```diff
@@ -0,0 +1,363 @@
+		optMaxStrongParents:          8,
+		optMaxLikedInsteadReferences: 8,
+		optMaxWeakReferences:         8,
+	}, opts, func(t *TipManager) {
+		t.optMaxLikedInsteadReferencesPerParent = t.optMaxLikedInsteadReferences / 2
+	})
+}
+
+// AddBlock adds a Block to the TipManager.
+func (t *TipManager) AddBlock(block *blocks.Block) {
+	newBlockMetadata := NewBlockMetadata(block)
+
+	if storage := t.metadataStorage(block.ID().Index()); storage != nil && storage.Set(block.ID(), newBlockMetadata) {
+		t.setupBlockMetadata(newBlockMetadata)
+	}
+}
+
+// SelectTips selects the references that should be used for block issuance.
+func (t *TipManager) SelectTips(amount int) (references model.ParentReferences) {
+	references = make(model.ParentReferences)
+
+	seenStrongTips := advancedset.New[iotago.BlockID]()
+	selectStrongTips := func(amount int) (strongTips []*TipMetadata) {
+		if amount > 0 {
+			for _, strongTip := range t.strongTipSet.RandomUniqueEntries(amount + seenStrongTips.Size()) {
+				if seenStrongTips.Add(strongTip.Block().ID()) {
+					if strongTips = append(strongTips, strongTip); len(strongTips) == amount {
+						return strongTips
+					}
+				}
+			}
+		}
+
+		return strongTips
+	}
+
+	_ = t.conflictDAG.ReadConsistent(func(conflictDAG conflictdag.ReadLockedConflictDAG[iotago.TransactionID, iotago.OutputID, booker.BlockVotePower]) error {
+		likedConflicts := advancedset.New[iotago.TransactionID]()
+
+		likedInsteadReferences := func(tipMetadata *TipMetadata) (references []iotago.BlockID, updatedLikedConflicts *advancedset.AdvancedSet[iotago.TransactionID], err error) {
+			necessaryReferences := make(map[iotago.TransactionID]iotago.BlockID)
+			if err = conflictDAG.LikedInstead(tipMetadata.Block().ConflictIDs()).ForEach(func(likedConflictID iotago.TransactionID) error {
+				transactionMetadata, exists := t.memPool.TransactionMetadata(likedConflictID)
+				if !exists {
+					return xerrors.Errorf("transaction required for liked instead reference (%s) not found in mem-pool", likedConflictID)
+				}
+
+				necessaryReferences[likedConflictID] = lo.First(transactionMetadata.Attachments())
+
+				return nil
+			}); err != nil {
+				return nil, nil, err
+			}
+
+			references, updatedLikedConflicts = make([]iotago.BlockID, 0), likedConflicts.Clone()
+			for conflictID, attachmentID := range necessaryReferences {
+				if updatedLikedConflicts.Add(conflictID) {
+					references = append(references, attachmentID)
+				}
+			}
+
+			if len(references) > t.optMaxLikedInsteadReferencesPerParent {
+				return nil, nil, xerrors.Errorf("too many liked instead references (%d) for block %s", len(references), tipMetadata.Block().ID())
+			}
+
+			return references, updatedLikedConflicts, nil
+		}
+
+		for strongTipCandidates := selectStrongTips(amount - len(references[model.StrongParentType])); len(strongTipCandidates) != 0; strongTipCandidates = selectStrongTips(amount - len(references[model.StrongParentType])) {
```

Comment:
During `stronTipCandidates` initialization, the `len(references[model.StrongParentType])` will always be 0 since this is the first time we touch `references` in this method.

Reply:
Yes, that is correct and intended, we initially assume that we will be able to reference all of the requested blocks (-0) but if we find some of them to have to be dropped, then we will try to fill up the missing ones in the next iteration (-n) where n is the ones we have found already - we essentially only request the missing candidates.

Reply:
But I mean that when we first call `selectStrongTips` it's always empty. Could we instead have the following? 
```suggestion 
for strongTipCandidates := selectStrongTips(amount); len(strongTipCandidates) != 0; strongTipCandidates = selectStrongTips(amount - len(references[model.StrongParentType])) {
```